### DISCUSSION TITLE
Should warn about Python 3.12 and Celery

### DISCUSSION THREAD

**Discussion by JohnRDOrazio** [WeblateOrg/weblate#10301] [General] [answered]
Seeing that Weblate 5 requires at least Python 3.9 which isn't available on Ubuntu 20.04, I went ahead and added the deadsnakes repository. I was then able to install a more recent version of Python, so I figured what the heck and I went and installed Python 3.12. I then updated my virtual environment against Python 3.12, fixed the symlink to point to Python 3.12, and then continued with the Weblate update. I passed first to Weblate 5.0.2 and then to Weblate 5.1.1. I had to force reinstall ffi libraries (only thing that worked was `pip install --force-reinstall --no-binary :all: cffi`). Migrations went well in both cases.

However, I am having a problem with celery. `weblate check --deploy` is giving me an error about celery not beating. When I check the celery logs, I see a number of warnings, in particular:
`WARNING/Beat] AttributeError: 'cached_property' object has no attribute 'lock'`

When doing a search for this error, I found [this stackoverflow issue](https://stackoverflow.com/questions/77268809/flask-celery-docker-compose-up-error-attributeerror-cached-property-obje) which led to this github celery issue:
[Assigning to cached_property is broken on Python 3.12](https://github.com/celery/kombu/issues/1804#top)

Basically, celery is currently incompatible with Python 3.12. Perhaps there should be a warning about this under the software requirements section: Python 3.9 or later but not Python 3.12! Until this issue is fixed.

**Comment by JohnRDOrazio** [comment]
I downgraded my virtual environment to Python 3.11, fixed the symlink for python, re-updated weblate, reinstalled uwsgi, force reinstalled ffi libraries. No migrations were needed this time, and after restarting the celery service and the uwsgi service I no longer get any warnings or errors from `weblate check --deploy`. My instance is back up and running with Python 3.11 and Weblate 5.1.1.

**Comment by nijel** [comment, marked as answer]
Documented in https://github.com/WeblateOrg/weblate/pull/10303, thanks for pointing this out.

**Reply by nijel** [reply]
Created https://github.com/WeblateOrg/weblate/issues/10361 to track this.
